File: 238-product-of-array-except-self/product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
const productExceptSelf = (nums) => {
  const output = new Array(nums.length);
  output.fill(1);

  let multiplier = 1;
  // prefix multiplication
  for (let i = 0; i < nums.length; i++) {
    output[i] = multiplier;
    multiplier *= nums[i];
  }
 
  multiplier = 1;
  for (let i = nums.length - 1; i >= 0; i--) {
    output[i] *= multiplier;
    multiplier *= nums[i];
  }
  return output;
};

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
    int[] output = new int[nums.length];
    int multiplier = 1;
    for (int i = 0; i < nums.length; i++) {
      output[i] = multiplier;
      multiplier = multiplier * nums[i];
    }
    multiplier = 1;
    for (int i = nums.length - 1; i >= 0; i--) {
      output[i] = output[i] * multiplier;
      multiplier = multiplier * nums[i];
    }
    return output;
    }
}